### ISSUE TITLE
New Onboarding: redirect new users to use-your-domain flow 

### ISSUE THREAD

**Issue by razvanpapadopol** [Automattic/wp-calypso#51137]
### What
Add the option to redirect to the existing `use-your-domain` flow on DomainStep in [New onboarding](https://wordpress.com/new/domains) for new (logged-out) users.

### Why
Allow users to create wp com sites using their own domain

### A/C
  - [ ] Show the signup modal when a new user is clicking the button
  - [ ] After successful signup/login, redirect to `use-your-domain` flow
  - [ ] Site creation after login/signup when the flow is complete and is not AnchorFM should continue to work
  - [ ] Continue to use query param when redirecting to Calypso Login and back on successful login

### How
Refactor [useOnLogin](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/gutenboarding/hooks/use-on-login.ts) and [useOnSignup](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/gutenboarding/hooks/use-on-signup.ts) hooks to stop creating the site by default at signup/login and do any action set by the Gutenboarding app.

Examples of such actions:
   - handle site creation when flow is complete
   - do nothing when the flow is AnchorFM 
   - redirect to `use-your-domain` when used [in DomainStep](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/gutenboarding/onboarding-block/domains/index.tsx#L150-L155)
   - start `/new` flow after Login / Signup

Basically we currently have `createSite` as the `action X` in the diagram below and we want it to be a configurable action.

<img width="1230" alt="Screenshot 2021-03-17 at 17 11 03" src="https://user-images.githubusercontent.com/14192054/111490711-ea655480-8743-11eb-997a-e8a45fe1d50d.png">
 
**Additional context**
PR that added this feature for existing users (already logged-in): https://github.com/Automattic/wp-calypso/pull/48692
Main issue: https://github.com/Automattic/wp-calypso/issues/44645

**Comment by ciampo** [comment]
Note to self on next steps, after some investigation today:

1. Refactor the hooks without changing functionality:
   - Extract logic for the `createSite` function out of the `useOnLogin` and `useOnSignup` hooks
   - Modify both hooks to accept a callback to execute. The callback will be a function that gets certain parameters (like userId and siteId). These parameters are going to be provided by the hooks
   - Pass the extracted `createSite` function as a callback to the hooks

2. Introduce the "Use a domain I own" button for logged out users
   - Show login/signup modal when clicking the button
   - Create a callback to redirect the user for the `useOnLogin` and `useOnSignup` hooks


Steps 1 and 2 should probably happen in 2 separate PR, to facilitate the review process

**Comment by razvanpapadopol** [comment]
Idea for handling Calypso Login success: pass the URL (`/start/domains/use-your-domain?source=${ window.location.href }`) to SignupForm and use it at https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/gutenboarding/components/signup-form/index.tsx#L193
